oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,575 @@
+// k/v pair.  It is referred to as u40 below.
+const INDEX_SIZE: usize = 5;
+// Size of an entry in the hash table.
+const ENTRY_SIZE: usize = INDEX_SIZE + 1;
+
+// The hash table consists of 2 vectors.  The first is a table of "entries", which is a packed
+// array of entries with the following layout:
+//
+//    hash_byte: u8,
+//    data_index: u40,  // A 5-byte index, so we support up to 1TiB of data.
+//
+// This is a "swiss" table meaning that instead of the usual "buckets" of entries, the entries are
+// collapsed into the hash table itself.  By adding a 1-byte hash, we can eliminate over 99.5% of
+// key comparisons to the wrong keys.  This is important for speed because comparing keys causes a
+// page fault usually, while comparing thie 1-byte hash is usually an L1 cache hit.  The number of
+// cache misses per lookup should be close to 2 for values found in the table, and 1 for values not
+// in the table.
+//
+// The storage for k/v pairs is chunked into 256 4GiB vectors:
+//
+//     data_chunks: Vec<Box([u8; N])>,  // N = 2^32
+//     used_chunks: usize,
+//
+// data_chunks starts empty, and as data is added to the table, we allocat another chunk.
+// The format for each k/v pair of data in each chunk is:
+//
+//     key_len: <compressed u30>
+//     key: [u8]
+//     value_len: <compressed u30>
+//     value: [u8]
+//
+// Another memory savings comes from using a swiss table with exactly a 60% load factor.  Swiss
+// tables require a significant portion of the table be empty, and normally as the table grows,
+// they dynamically double the size.  At least when using the alloc crate, this resize causes a new
+// table to be allocated at 2X the size, data is copied from the old table to the new, and the old
+// table is freed.  This results in a 3X expansion in memory usage vs the prior array size.  In
+// contrast, this scheme has only 40% of the table empty and is never resized.
+//
+// This reduce scheme currently limits the number of entries to 2^32.
+
+const CHUNK_BITS: usize = 21;
+const CHUNK_SIZE: usize = 1 << CHUNK_BITS;
+const CHUNK_MASK: usize = CHUNK_SIZE - 1;
+
+#[derive(Default)]
+pub struct LookupHtbl {
+    table: Vec<u8>, // Contains fixed-size entries with index of k/v data in data_chunks.
+    data_chunks: Vec<Box<[u8]>>,
+    max_entries: usize,       // The number at which we must grow the table.
+    allocated_entries: usize, // This is self.table.len() * ENTRY_SIZE.
+    used_data: usize,
+    used_entries: usize,
+}
+
+enum LookupResult {
+    Found(usize),        // (data_index: u40)
+    NotFound(usize, u8), // (table_index: u40, hash_byte: u8)
+}
+
+impl LookupHtbl {
+    // Set the initial size of self.table.  For best speed and memory use, this should be as large
+    // as the number of key/value entries that will be loaded.
+    pub fn reserve(&mut self, max_entries: usize) {
+        assert!(self.used_entries == 0 && self.used_data == 0);
```

Comment:
Document that reserve is only supposed to be called once.

Alternatively:
a) could we return an error?
b) maybe turn this into a factory method (similar to [`HashMap::with_capacity`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.with_capacity))

Reply:
Done.

Reply:
Returning an error would be inconsistent with HashMap.  For now, I've just added the comment.